### ISSUE TITLE
pp/vagrant-cloud: large files are read in to memory

### ISSUE THREAD

**Issue by beenje** [hashicorp/packer#5004]
Hi,

I'm trying to switch from the `atlas` post processor to `vagrant-cloud` to upload a vagrant box.

This was working fine:
```json
  "post-processors": [
    [
      {
        "type": "vagrant",
        "vagrantfile_template": "templates/vagrantfile.tpl",
        "keep_input_artifact": true,
        "output": "build/esss-devenv-7.3.box"
      },
      {
        "type": "atlas",
        "artifact": "esss/devenv-7.3",
        "artifact_type": "vagrant.box",
        "metadata": {
          "description": "{{user `description`}}",
          "provider": "virtualbox",
          "version": "{{user `devenv_version`}}"
        }
      }
    ]
  ]
```

This doesn't work:
```json
  "post-processors": [
    [
      {
        "type": "vagrant",
        "vagrantfile_template": "templates/vagrantfile.tpl",
        "output": "build/esss-devenv-7.3.box"
      },
      {
        "type": "vagrant-cloud",
        "access_token": "{{user `vagrantcloud_token`}}",
        "box_tag": "esss/devenv-7.3",
        "no_release": "true",
        "version": "{{user `devenv_version`}}",
        "version_description": "{{user `description`}}"
      }
    ]
  ]
```

It raises the error:
* Post-processor failed: unexpected EOF

See the gist url below for the full debug log.

- Packer version: Packer v1.0.0
- Host platform: CentOS Linux release 7.3.1611
- Debug log: https://gist.github.com/beenje/71a711b0587d18714abebd417f691158
- Pakcer template used: https://gist.github.com/beenje/568aabe7881c649ca7baec4514d89e79

**Comment by rickard-von-essen** [comment]
Clould you fix your json in the OP, thanks.

**Comment by beenje** [comment]
Fixed. Sorry.
You have the full packer template with the atlas post processor here if needed: https://gist.github.com/beenje/5a50d96dc2a719e27593a5f2188fc27e

**Comment by mwhooker** [comment]
here's the issue

```
2017/06/12 13:13:47 packer: runtime: out of memory: cannot allocate 2147483648-byte block (2152726528 in use)
2017/06/12 13:13:47 packer: fatal error: out of memory
```

so we're probably reading something in to memory when we shouldn't. Will investigate more

**Comment by mwhooker** [comment]
@beenje hi, any chance you can test my patch, #5005? I can give you a binary if you're not able to build packer.

**Comment by beenje** [comment]
@mwhooked, I just tested your patch and it worked. Thanks!
The full debug log is here: https://gist.github.com/beenje/2b3e868c961403df004b6425f5917450

Is it normal to have 3 attempts for the upload?
```
2017/06/12 22:06:23 ui:     virtualbox-ovf (vagrant-cloud): Box upload prepared with token
2017/06/12 22:06:23 ui: ==> virtualbox-ovf (vagrant-cloud): Uploading box: build/esss-devenv-7.3.box
2017/06/12 22:06:23 ui:     virtualbox-ovf (vagrant-cloud): Depending on your internet connection and the size of the box,
    virtualbox-ovf (vagrant-cloud): this may take some time
2017/06/12 22:06:23 packer: 2017/06/12 22:06:23 Post-Processor Vagrant Cloud API Upload: build/esss-devenv-7.3.box https://archivist.hashicorp.com/v1/object/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJrZXkiOiJib3hlcy85MjY2M2FiZi1hMjRjLTQ1YWMtYWY4YS1iNDY0MDg2OTNiNTUiLCJtb2RlIjoidyIsImV4cGlyZSI6MTQ5NzMyNjc4M30.RrLTuJGu17ws5S2DHM872ABBk7TLTaax3O03MdQm_rM
2017/06/12 22:07:17 packer: 2017/06/12 22:07:17 Post-Processor Vagrant Cloud Upload Response:
2017/06/12 22:07:17 packer:
2017/06/12 22:07:17 packer: &{Status:200 OK StatusCode:200 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Access-Control-Allow-Methods:[GET, PUT, PATCH, HEAD, OPTIONS] Access-Control-Allow-Origin:[https://atlas.hashicorp.com] Content-Type:[text/plain; charset=utf-8] Connection:[keep-alive] Access-Control-Allow-Headers:[Accept-Encoding, Connection, User-Agent, X-Forwarded-For, X-Forwarded-Port, X-Forwarded-Proto] Access-Control-Expose-Headers:[Location] Date:[Mon, 12 Jun 2017 20:07:17 GMT] Content-Length:[0]] Body:{} ContentLength:0 TransferEncoding:[] Close:false Uncompressed:false Trailer:map[] Request:0xc42025c600 TLS:0xc42044a790}
2017/06/12 22:07:17 ui:     virtualbox-ovf (vagrant-cloud): Uploading box, attempt 2
2017/06/12 22:07:17 packer: 2017/06/12 22:07:17 Post-Processor Vagrant Cloud API Upload: build/esss-devenv-7.3.box https://archivist.hashicorp.com/v1/object/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJrZXkiOiJib3hlcy85MjY2M2FiZi1hMjRjLTQ1YWMtYWY4YS1iNDY0MDg2OTNiNTUiLCJtb2RlIjoidyIsImV4cGlyZSI6MTQ5NzMyNjc4M30.RrLTuJGu17ws5S2DHM872ABBk7TLTaax3O03MdQm_rM
2017/06/12 22:08:15 packer: 2017/06/12 22:08:15 Post-Processor Vagrant Cloud Upload Response:
2017/06/12 22:08:15 packer:
2017/06/12 22:08:15 packer: &{Status:200 OK StatusCode:200 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Access-Control-Allow-Headers:[Accept-Encoding, Connection, User-Agent, X-Forwarded-For, X-Forwarded-Port, X-Forwarded-Proto] Access-Control-Allow-Methods:[GET, PUT, PATCH, HEAD, OPTIONS] Content-Type:[text/plain; charset=utf-8] Content-Length:[0] Access-Control-Allow-Origin:[https://atlas.hashicorp.com] Access-Control-Expose-Headers:[Location] Date:[Mon, 12 Jun 2017 20:08:15 GMT] Connection:[keep-alive]] Body:{} ContentLength:0 TransferEncoding:[] Close:false Uncompressed:false Trailer:map[] Request:0xc42000a700 TLS:0xc42044a790}
2017/06/12 22:08:15 ui:     virtualbox-ovf (vagrant-cloud): Uploading box, attempt 3
2017/06/12 22:08:15 packer: 2017/06/12 22:08:15 Post-Processor Vagrant Cloud API Upload: build/esss-devenv-7.3.box https://archivist.hashicorp.com/v1/object/eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJrZXkiOiJib3hlcy85MjY2M2FiZi1hMjRjLTQ1YWMtYWY4YS1iNDY0MDg2OTNiNTUiLCJtb2RlIjoidyIsImV4cGlyZSI6MTQ5NzMyNjc4M30.RrLTuJGu17ws5S2DHM872ABBk7TLTaax3O03MdQm_rM
2017/06/12 22:09:05 packer: 2017/06/12 22:09:05 Post-Processor Vagrant Cloud Upload Response:
2017/06/12 22:09:05 packer:
2017/06/12 22:09:05 packer: &{Status:200 OK StatusCode:200 Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[Access-Control-Allow-Methods:[GET, PUT, PATCH, HEAD, OPTIONS] Access-Control-Expose-Headers:[Location] Date:[Mon, 12 Jun 2017 20:09:05 GMT] Access-Control-Allow-Headers:[Accept-Encoding, Connection, User-Agent, X-Forwarded-For, X-Forwarded-Port, X-Forwarded-Proto] Access-Control-Allow-Origin:[https://atlas.hashicorp.com] Content-Type:[text/plain; charset=utf-8] Content-Length:[0] Connection:[keep-alive]] Body:{} ContentLength:0 TransferEncoding:[] Close:false Uncompressed:false Trailer:map[] Request:0xc420070400 TLS:0xc42044a790}
2017/06/12 22:09:05 ui:     virtualbox-ovf (vagrant-cloud): Box successfully uploaded
```

**Comment by mwhooker** [comment]
I'm not sure right now. We've had another report of that over here though. https://github.com/hashicorp/packer/issues/4973#issuecomment-307446717

We're looking in to it

**Comment by mwhooker** [comment]
@beenje the multiple upload issue should be fixed in #5008

**Comment by beenje** [comment]
@mwhooker thanks for the feedback!
